File: Striver/Trees/Q15.py
# zigzag traversal of a binary tree
from Node import Node

def zigzag(root):
    ans = []
    if root == None:
        return ans
    q = []
    q.append(root)
    flag = 0
    while len(q) > 0:
        size = len(q)
        temp = []
        for i in range(size):
            node = q.pop(0)
            temp.append(node.data)
            if node.left != None:
                q.append(node.left)
            if node.right != None:
                q.append(node.right)
        if flag == 0:
            ans.append(temp)
            flag = 1
        else:
            ans.append(temp[::-1])
            flag = 0
    return ans

root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(7)
root.left.right = Node(6)
root.right.left = Node(5)
root.right.right = Node(4)

print(zigzag(root))

File: Striver/Trees/Q15.java
package Striver.Trees;

import java.util.*;

public class Q15{
    // ZigZag traversal of a binary tree

    public List<List<Integer>> zigzagLevelOrder(Node root) {
        List<List<Integer>> ans = new ArrayList<>();
        if(root == null)return ans;
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        boolean flag = true;
        while(!q.isEmpty()){
            int size = q.size();
            List<Integer> temp = new ArrayList<>();
            while(size-- > 0){
                Node curr = q.poll();
                temp.add(curr.data);
                if(curr.left != null)q.add(curr.left);
                if(curr.right != null)q.add(curr.right);
            }
            if(flag){
                ans.add(temp);
                flag = false;
            }else{
                Collections.reverse(temp);
                ans.add(temp);
                flag = true;
            }
        }
        return ans;
    }

    public static void main(String[] args) {
        Node root = new Node(3);
        root.left = new Node(9);root.right = new Node(20);
        root.right.left = new Node(15);root.right.right = new Node(7);
        Q15 q = new Q15();
        List<List<Integer>> ans = q.zigzagLevelOrder(root);
        for(List<Integer> i : ans){
            for(int j : i){
                System.out.print(j+" ");
            }
            System.out.println();
        }
    }

}